torch_geometric/nn/models/graph_unet.py

```diff
@@ -0,0 +1,112 @@
+
+    Args:
+        channels (int): Size of each sample.
+        depth (int): The depth of the U-Net architecture.
+        pool_ratios (float or [float], optional): Graph pooling ratio for each
+            depth. (default: :obj:`0.5`)
+        sum_res (bool, optional): If set to :obj:`False`, will use
+            concatenation for integration of skip connections instead
+            summation. (default: :obj:`True`)
+        act (torch.nn.functional, optional): The nonlinearity to use.
+            (default: :obj:`torch.nn.functional.relu`)
+    """
+
+    def __init__(self, channels, depth, pool_ratios=0.5, sum_res=True,
+                 act=F.relu):
+        super(GraphUNet, self).__init__()
+
+        self.channels = channels
+        self.depth = depth
+        self.pool_ratios = repeat(pool_ratios, depth)
+
+        self.act = act
+        self.sum_res = sum_res
+
+        self.pools = torch.nn.ModuleList()
+        self.down_convs = torch.nn.ModuleList()
+        for i in range(depth):
+            self.pools.append(TopKPooling(channels, self.pool_ratios[i]))
+            self.down_convs.append(GCNConv(channels, channels, improved=True))
+        self.up_convs = torch.nn.ModuleList()
+        for i in range(depth - 1):
+            in_channels = channels if sum_res else 2 * channels
+            self.up_convs.append(GCNConv(in_channels, channels, improved=True))
+
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        for pool in self.pools:
+            pool.reset_parameters()
+        for conv in self.down_convs:
+            conv.reset_parameters()
+        for conv in self.up_convs:
+            conv.reset_parameters()
+
+    def forward(self, x, edge_index, batch=None):
+        if batch is None:
+            batch = edge_index.new_zeros(x.size(0))
+        edge_weight = x.new_ones(edge_index.size(1))
+
+        xs = [x]
+        edge_indices = [edge_index]
+        edge_weights = [edge_weight]
+        perms = []
+
+        for i in range(self.depth):
+            edge_index, edge_weight = self.augment_adj(edge_index, edge_weight,
+                                                       x.size(0))
+            x, edge_index, edge_weight, batch, perm, _ = self.pools[i](
+                x, edge_index, edge_weight, batch)
+            x = self.act(self.down_convs[i](x, edge_index, edge_weight))
+
+            if i < self.depth - 1:
+                xs += [x]
+                edge_indices += [edge_index]
+                edge_weights += [edge_weight]
+            perms += [perm]
+
+        for i in range(self.depth):
+            j = self.depth - 1 - i
+
+            res = xs[j]
+            edge_index = edge_indices[j]
+            edge_weight = edge_weights[j]
+            perm = perms[j]
+
+            up = torch.zeros_like(res)
+            up[perm] = x
+            x = res + up if self.sum_res else torch.cat((res, up), dim=-1)
+
+            if i < self.depth - 1:
```

Comment:
I don't think this is correct: you need to finish on a convolution! Right now if you are concatenaing, the output will be of dim*2 and not dim.